src/advanced_inv.cpp

```diff
@@ -1365,111 +1428,99 @@ void advanced_inventory::display()
+                        std::copy(mv.begin(), mv.end(), std::back_inserter(moving_items));
+                    }
+                    int items_left = 0, moved = 0;
+                    for(auto &elem : moving_items) {
+                        assert(!elem.is_null());
+                        items_left = add_item(destarea, elem);
+                        if(items_left > 0) {
+                            // chargeback the items if adding them failed
+                            add_item(srcarea, elem, items_left);
+                        } else {
+                            ++moved;
                         }
                     }
+                    if(moved == 0) {
+                        continue;
+                    }
                 }
             } else {
                 // from map/vehicle: add the item to the destination, if that worked,
                 // remove it from the source, else continue.
-                // TODO: move partial stack with items that are not counted by charges
-                item new_item( *sitem->it );
-                if( by_charges ) {
-                    new_item.charges = amount_to_move;
-                }
-                if( !add_item( destarea, new_item ) ) {
-                    // make sure to redraw the info window, so 
-                    // the user can see why it wasn't equipped
-                    if(destarea == AIM_WORN) {
-                        redraw = 1;
+                // create a new copy of the old item being manipulated
+                item new_item(*sitem->items.front());
+                // add the item, and note any items that might be leftover
+                int items_left = add_item(destarea, new_item, (by_charges) ? 1 : amount_to_move);
+                // only remove item or charges if the add succeeded
+                if(items_left == 0) {
+                    if(by_charges) {
+                        // set the new item's charge amount
+                        new_item.charges = amount_to_move;
+                        // `amount_to_move' will be `true' if the item needs to be removed
+                        amount_to_move = sitem->items.front()->reduce_charges(amount_to_move);
                     }
-                    continue;
+                    // take into account only the items _left over_ from transaction
+                    items_left -= remove_item(*sitem, amount_to_move);
                 }
-
-                if( by_charges && amount_to_move < sitem->it->charges ) {
-                    sitem->it->charges -= amount_to_move;
-                } else {
-                    remove_item( *sitem );
+                if(items_left > 0) {
+                    const char *msg = nullptr;
+                    if(by_charges) {
+                        msg = _("Only moved %d of %d charges.");
+                    } else {
+                        msg = _("Only moved %d of %d items.");
+                        remove_item(*sitem, items_left);
```

Comment:
Wait, why are you removing the items that have *not* been moved (that are left)?